app/lib/search/search_service.dart

```diff
@@ -444,33 +455,52 @@ class TagsPredicate {
         tag = tag.substring(1);
       }
       if (required) {
-        requiredTags.add(tag);
+        p._values[tag] = true;
       } else {
-        negatedTags.add(tag);
+        p._values[tag] = false;
       }
     }
-    return TagsPredicate(requiredTags: requiredTags, negatedTags: negatedTags);
+    return p;
+  }
+
+  /// Appends [requiredTags] and [prohibitedTags] to the current set of tags,
+  /// and returns a new [TagsPredicate] instance.
+  /// The appended flags take precedence over the current ones.
+  TagsPredicate append(
+      {List<String> requiredTags, List<String> prohibitedTags}) {
+    final p = TagsPredicate();
+    p._values.addAll(_values);
+    requiredTags?.forEach((tag) => _values[tag] = true);
+    prohibitedTags?.forEach((tag) => _values[tag] = false);
+    return p;
+  }
+
+  /// Appends [other] predicate to the current set of tags, and returns a new
+  /// [TagsPredicate] instance.
+  /// The appended flags take precedence over the current ones.
+  TagsPredicate appendPredicate(TagsPredicate other) {
+    final p = TagsPredicate();
+    p._values.addAll(_values);
+    p._values.addAll(other._values);
+    return p;
   }
 
   /// Evaluate this predicate against the list of supplied [tags].
   /// Returns true if the predicate matches the [tags], false otherwise.
   bool evaluate(List<String> tags) {
```

Comment:
nit: (out of scope) Maybe we should rename this `match`, `isMatch`, `hasMatch` or `satisfied`..